extract `model.pushObjects` to a method

asross opened the pull request:
...to allow overrides in the case when more specialized munging/manipulation might be required.

For one part of our app that needs infinite scroll, we have a slightly non-traditional API that requires a more complicated merging of our initial data and subsequent requests. This extremely minor change makes it much easier for us to use `ember-infinity` in that case.

kellyselden commented on addon/mixins/route.js:
```diff
@@ -217,6 +217,10 @@ export default Ember.Mixin.create({
     return false;
   },
 
+  _updateModel(model, infinityModel) {
```
If the purpose is overriding, wouldn't you want to make it public?

hhff replied:
Yeah agree!  I was under the impression that the "pushObjects" to model content no longer works in new versions of Ember Data.  What version of Ember Data are you using?

asross replied:
Will un-underscore the method name. I'm not sure about the compatibility of `pushObjects` with the latest Ember data, but that doesn't seem relevant to this change (since I'm just moving code around).

kellyselden commented:
I would suggest a README change to help others with this. Then squash commits and LGTM.

hhff commented on addon/mixins/route.js:
```diff
@@ -217,6 +217,10 @@ export default Ember.Mixin.create({
     return false;
   },
 
+  updateModel(model, infinityModel) {
```
We also don't need to pass `infinityModel` here, we can get it with `this.get('_modelPath')`.

Can we also rename this method to `addObjects`, and the first arg to `objects`?  Want to be explicit that we aren't updating or changing the infinityModel in place, we're just adding to it.

Also - a small note in the README would be great!

kellyselden replied:
@hhff I think the point of the `infinityModel` param is to use it when overriding without accessing anything private.

hhff replied:
Its fine to use private methods inside a public method?

kellyselden replied:
Are you saying that when overriding, people should use `_modelPath` to get the `infinityModel`?

hhff replied:
no, im saying we don't even need the 2nd arg, cause we can retrieve it from the route

kellyselden replied:
Hmm. I think you know better than me in this area, I'm going to bow out.

hhff replied:
I may be missing something!  The purpose of this method is just for conveniently pushing arbitary objects into the infinityModel from the route, right @asross ?

asross replied:
Sorry for being so late to reply! I'd be in favor of keeping `infinityModel` as a parameter because we don't necessarily want users of this addon to have to think deeply about implementation details of `ember-infinity`, instead we just want them to specify how the response from the server (`infinityModel`) should update the route's data (`model`). So from an addon-overrider user experience point of view, I think it's slightly nicer to have it be an argument.

Another argument-argument is that you might change the private variable name later, but you don't necessarily want users of the addon to have to change their code.

But, from a general software point of view, it does make more sense to omit the second argument since it _is_ superfluous. And if you're overriding this method, even though it's public, you maybe know what you're in for.

In any case, I'm happy with either, the main thing is to have a separate method. I'll add something to the README to this effect and whichever you think is better @hhff  we will go with.

asross replied:
Actually, I'm wondering if we should rename this method to something more specific to `ember-infinity`? It's conceivable that an app could already have a route method called `updateModel` that's unrelated to this addon.

asross replied:
I changed it to `updateInfinityModel` to avoid name collision, hope that's alright and sorry for inadvertently hiding this chain of comments.

hhff commented on addon/mixins/route.js:
```diff
@@ -217,6 +217,10 @@ export default Ember.Mixin.create({
+  updateInfinityModel(model, infinityModel) {
+    model.pushObjects(infinityModel.get('content'));
+  },
+
```
I'm still finding this confusing.  the `infinityModel` should always refer to the route's model, ie, the model stored at `_modelPath`.

I think we should do

```js
updateInfinityModel(collection, infinityModel) {
    if (!infinityModel) {
      infinityModel = this.get('_modelPath');
    }
    return infinityModel.pushObjects(collection.get('content'));
}
```

This would imply we'd want to change the server response in the promise above from `infinityModel` to `collection`, and then line 191 to: `this.updateInfinityModel(collection, model);`

I think the wording `infinityModel` should canonically always refer to the object at `route.get('_modelPath')` and nothing else.

This also means that developers can also use this function like so:

```js
actions: {
  pushHughIntoInfinityModel() [
    var updatedInfinityModel = this.updateInfinityModel([
      { id: 1, name: "Hugh Francis" }
    ]);
    console.log(updatedInfinityModel);
  }
}
```

...which is something some other devs have asked about.

asross replied:
:+1:, I really like that. Will make that change and potentially update the README too.

hhff replied:
Psyched!  sorry for the nitpicks, just think this is an important addition to the API and I wanna get it right xo

asross commented:
I updated the method signature and also renamed some variables so that `infinityModel` consistently refers to the route's model rather than the new objects to be added to it. Let me know if the README example which I c/ped from you is ok :grin:

hhff commented on addon/mixins/route.js:
```diff
@@ -217,6 +216,19 @@ export default Ember.Mixin.create({
+
+    return infinityModel.pushObjects(newObjects);
+  },
+
```
Ahhh now the problem is that we've pulled the `newObjects.get('content')` back into the load method.  That means, if someone is using pure arrays without prototype extensions, there won't be a good place for them to override the `get('content')` part of the push.

I like this idea because it exposes the method for actually extracting the data from the collection that is pushed into the `infinityModel`, giving developers a public place to change that logic.

asross replied:
Hmm, but doesn't that mean you would no longer be able to do `updateInfinityModel` with a literal array?

asross replied:
I think the example you gave wouldn't be possible if we did the `get('content')` inside this method (without an override).

hhff replied:
It would work for the majority of developers, unless they've explicitly turned off prototype extensions, and if they have, they're probably going to be aware of that stuff.

asross replied:
Ok, sounds good. Will make the change.

asross replied:
Done, apologies for the delay.

hhff commented:
this is perfect @asross - can you put up a simple test?  Once we've got some test coverage this is good to merge

kellyselden commented on README.md:
````diff
@@ -111,6 +111,19 @@ and ember-infinity will be set up to parse the total number of pages from a JSON
+
+```js
+actions: {
+  pushHughIntoInfinityModel() [
````
nice name choice :)

asross commented on tests/unit/mixins/route-test.js:
```diff
@@ -308,3 +308,70 @@ test('it uses overridden params when loading more data', assert => {
+    content: Ember.A([
+      { id: 3, title: 'Tender Is the Night' }
+    ])
+  });
```
This was a little bit laborious to construct... I might be wrong, but I don't think manual invocations of `updateInfinityModel` as specified in your example @hhff will work even with prototype extensions. However, it's still at least possible, and maybe it's nice to have the flexibility of being able to access meta parameters in `updateInfinityModel`.

So my default plan will be to keep this and update the README so that instead of
```js
var updatedInfinityModel = this.updateInfinityModel([
  { id: 1, name: "Hugh Francis" }
]);
```
we have:
```js
var updatedInfinityModel = this.updateInfinityModel(Ember.ArrayProxy.create({
  content: Ember.A([
    { id: 1, name: "Hugh Francis" }
  ])
});
```
Let me know how that sounds or if I just am misunderstanding prototype extensions!

asross replied:
Ok, reading up on it a bit more, I was definitely misunderstanding prototype extensions and some aspects of this addon. Prototype extensions are disabled locally in this project but we expect that most consumers of the addon will have them enabled. So hopefully the example above should work?

The only thing I'm not sure about now is whether an ES6 or ES5 prototype-extended Array will respond properly to `.get('content')`. But I think I don't need to update the README as drastically as I thought. This has been a nice learning experience.

hhff replied:
Yeah!  I think this is ready...!  I think if people want to use non-prototype extension'd collections, we'll just recommend they override the hook to account for that case.

hhff commented:
Can u give it a quick squash?

asross commented:
> Can u give it a quick squash?

Yep, done!

hhff commented:
:+1:

kellyselden commented:
Nice work.

asross commented:
Thanks!
